File: leetcode-cn/sword2offer/000剑指0_Offer_15._二进制中1的个数.py
'''
执行用时：44 ms, 在所有 Python3 提交中击败了46.65% 的用户
内存消耗：15 MB, 在所有 Python3 提交中击败了5.24% 的用户
'''

class Solution:
    def hammingWeight(self, n: int) -> int:
        return bin(n).count('1')

'''
执行用时：36 ms, 在所有 Python3 提交中击败了88.19% 的用户
内存消耗：14.8 MB, 在所有 Python3 提交中击败了43.86% 的用户
'''
class Solution:
    def hammingWeight(self, n: int) -> int:
        cnt = 0
        while n:
            n &= (n - 1) 
            cnt += 1
        return cnt


'''
执行用时：60 ms, 在所有 Python3 提交中击败了9.27% 的用户
内存消耗：15 MB, 在所有 Python3 提交中击败了5.24% 的用户
'''
class Solution:
    def hammingWeight(self, n: int) -> int:
        return 0 if n <= 0 else 1 + self.hammingWeight(n & (n - 1))
                

File: leetcode-cn/sword2offer/000剑指0_Offer_15._二进制中1的个数.cpp
/**
执行用时：0 ms, 在所有 C++ 提交中击败了100.00% 的用户
内存消耗：6 MB, 在所有 C++ 提交中击败了6.15% 的用户
 * **/


class Solution {
public:
    int hammingWeight(uint32_t n) {
        return __builtin_popcount(n);
    }
}; 



/*
执行用时：0 ms, 在所有 C++ 提交中击败了100.00% 的用户
内存消耗：5.9 MB, 在所有 C++ 提交中击败了12.89% 的用户
*/

class Solution {
public:
    int hammingWeight(uint32_t n) {
        int cnt = 0;
        for (int i = 0; i < 32; ++i)
        {
            if (n & (1 << i))
                cnt ++;
        }
        return cnt;
    }
};






/*
C

执行用时：0 ms, 在所有 C 提交中击败了100.00% 的用户
内存消耗：5.4 MB, 在所有 C 提交中击败了20.02% 的用户
*/
int hammingWeight(uint32_t n) {
    int cnt = 0;
    for (int i = 0; i < 32; ++i) {
        if (n & (1u << i)) {
            cnt ++;
        }
    }
    return cnt;
}



 
